### PR TITLE
St/evm 330

### PR DESCRIPTION
**Pull request by saratnt**

### CODE REVIEW

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 1]
```diff
@@ -149,8 +142,10 @@ class MempoolMap(stateReader: AccountStateReader) extends ScorexLogging {
    */
   def takeExecutableTxs(limit: Int): Iterable[SidechainTypes#SCAT] = {
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:143]: Simple case: we have 1M txs in the mempool
Forger will take all 1M sorted elements and than will take only the first ~1400 valid txs.

An improvement here can be to have a lazy data structure.
We may put it into backlog

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 2]
```diff
@@ -175,7 +170,197 @@ class MempoolMap(stateReader: AccountStateReader) extends ScorexLogging {
   }
 
   def canPayHigherFee(newTx: SidechainTypes#SCAT, oldTx: SidechainTypes#SCAT): Boolean = {
-     (newTx.getMaxFeePerGas.compareTo(oldTx.getMaxFeePerGas) > 0) &&
-      (newTx.getMaxPriorityFeePerGas.compareTo(oldTx.getMaxPriorityFeePerGas) > 0)
+    (newTx.getMaxFeePerGas.compareTo(oldTx.getMaxFeePerGas) > 0) &&
+    (newTx.getMaxPriorityFeePerGas.compareTo(oldTx.getMaxPriorityFeePerGas) > 0)
+  }
+
+  def updateMemPool(rejectedBlocks: Seq[AccountBlock], appliedBlocks: Seq[AccountBlock]): Unit = {
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:177]: As for me, comments for every logical block here and sub-methods are very useful.
A lot of modifications are done considering some assumptions.

File: sdk/src/test/scala/com/horizen/account/performance/AccountSidechainNodeViewHolderPerfTest.scala [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+      }
+    }
+
+    wallet = mock[AccountWallet]
+    Mockito.when(wallet.scanOffchain(ArgumentMatchers.any[SidechainTypes#SCAT])).thenReturn(wallet)
+  }
+
+  @Test
+  @Ignore
+  def txModifyTest(): Unit = {
```
Comment (i-Alex) [inline sdk/src/test/scala/com/horizen/account/performance/AccountSidechainNodeViewHolderPerfTest.scala:75]: Same here, can be good to add explanation comments to easily understand the flow and what we are measuring here without deep diving into the code.

**Review by i-Alex — CHANGES_REQUESTED** [review]
Looks good!
Minor points only

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala (hunk 1 above)
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:143]: Ok, probably I will see it in the recurrent PR https://github.com/HorizenOfficial/Sidechains-SDK/pull/593

File: sdk/src/main/scala/com/horizen/account/mempool/AccountMemoryPool.scala [hunk 4]
```diff
@@ -134,10 +120,20 @@ class AccountMemoryPool(
...
-    new AccountMemoryPool(new MempoolMap(stateReader), stateReader)
+  def createEmptyMempool(stateReaderProvider: AccountStateReaderProvider): AccountMemoryPool = {
+    new AccountMemoryPool(new MempoolMap(stateReaderProvider), stateReaderProvider)
   }
 }
+
+trait AccountStateReaderProvider {
+  def getAccountStateReader(): AccountStateReader
+
+}
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/AccountMemoryPool.scala]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 5]
```diff
@@ -149,8 +142,10 @@ class MempoolMap(stateReader: AccountStateReader) extends ScorexLogging {
    */
   def takeExecutableTxs(limit: Int): Iterable[SidechainTypes#SCAT] = {
 
+    val baseFee = stateReaderProvider.getAccountStateReader().baseFee
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:146]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 6]
```diff
@@ -149,8 +142,10 @@ class MempoolMap(stateReader: AccountStateReader) extends ScorexLogging {
...
+
     def txOrder(tx: SidechainTypes#SCAT) = {
-      tx.getMaxFeePerGas.subtract(stateReader.baseFee).min(tx.getMaxPriorityFeePerGas)
+      tx.getMaxFeePerGas.subtract(baseFee).min(tx.getMaxPriorityFeePerGas)
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:149]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:146]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala (hunk 1 above)
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:144]: Correct, this issue was addressed in PR #593

File: sdk/src/test/scala/com/horizen/account/performance/AccountSidechainNodeViewHolderPerfTest.scala (hunk 3 above)
Comment (saratnt) [reply sdk/src/test/scala/com/horizen/account/performance/AccountSidechainNodeViewHolderPerfTest.scala:85]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala (hunk 2 above)
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:181]: Done

**Pull request merged**